### PR TITLE
REP-4867 Character Encoding Support for Valkyrie Culling

### PR DESCRIPTION
**Pull request by dmnjohns**

### CODE REVIEW

**Comment by adrianjgeorge** [comment]
Just comments about the docs really.

File: repose-aggregator/src/docs/asciidoc/filters/valkyrie-authorization.adoc [hunk 1]
```diff
@@ -2,51 +2,224 @@
...
+[NOTE]
+====
+Currently, only JSON bodies are supported.
+If you would like to request XML support, please <<../contact-us.adoc#,contact us>>!
+====
+* **Account Permission to Role Translation**:
+Takes the user's account level permissions and translates their names into the standard `X-Roles` header allowing any of the standard means of RBAC to be used on the request.
+This includes translating the user's device permission for the device specified in the `X-Device-Id` header (when the request contains one).
+
+While the authorization function is always performed, the other functions may be enabled or disabled as desired.
```
Comment (dmnjohns) [inline repose-aggregator/src/docs/asciidoc/filters/valkyrie-authorization.adoc]: It seems like it is? Well, you can get around it somewhat with the `pre-authorized-roles` feature, but from what I can tell, we always fetch the permissions, and assuming that succeeds, check to see if those permissions allow access to the device in question.

https://github.com/rackerlabs/repose/blob/master/repose-aggregator/components/filters/valkyrie-authorization-filter/src/main/scala/org/openrepose/filters/valkyrieauthorization/ValkyrieAuthorizationFilter.scala#L284-L287

File: repose-aggregator/src/docs/asciidoc/filters/valkyrie-authorization.adoc [hunk 2]
```diff
@@ -2,51 +2,224 @@
...
+* `X-Tenant-Id`
+* `X-Device-Id` (Only required for per-device authorization)
 
 === Required preceding filters
-???.
+While there are no preceding filters that are strictly required, the following filters may be useful in certain situations:
+
+* <<header-normalization.adoc#, Header Normalization filter>> - Can be used to ensure users are not able to set their own `X-Device-Id` header (which could elevate privileges).
+* <<keystone-v2.adoc#, Keystone v2 filter>> - Can be used to enhance the request with the the `X-Contact-Id` and `X-Tenant-Id` headers.
+* <<api-validator.adoc#, API Validator filter>> - Can be used to capture the `X-Device-Id` header from the URI.
```
Comment (dmnjohns) [inline repose-aggregator/src/docs/asciidoc/filters/valkyrie-authorization.adoc]: Oops. Will change.

File: repose-aggregator/src/docs/asciidoc/filters/valkyrie-authorization.adoc [hunk 3]
```diff
@@ -2,51 +2,224 @@
...
+* `X-Roles` - Only created if this filter is configured to `translate-permissions-to-roles`.
+The value of this header will be the names of the user's permissions.
 
 === Request body changes
-???.
+This filter does not modify the request body.
 
 === Recommended follow-on (succeeding) filters
-???.
+This filter is not strictly required by any other filters.
```
Comment (dmnjohns) [inline repose-aggregator/src/docs/asciidoc/filters/valkyrie-authorization.adoc]: I'll add those.

File: repose-aggregator/src/docs/asciidoc/filters/valkyrie-authorization.adoc [hunk 4]
```diff
@@ -2,51 +2,224 @@
...
+If not configured, the default connection pool is used.
+<5> Specifies whether or not to verify that the `X-Tenant-Id` header starts with `hybrid:`.
+If disabled (`false`, the default), this filter will verify that the `X-Tenant-Id` header starts with `hybrid:` (indicating a dedicated tenant) before making a request to Valkyrie.
+If the header does not start with `hybrid:` (indicating a non-dedicated tenant), the request is immediately rejected.
+If enabled (`true`), the filter will be skipped for non-dedicated tenants allowing processing to continue.
+<6> Specifies whether or not to send a failing response when an invalid state is reach.
+If present, the filter will not send a failing response.
+Instead, it will add the data relating to the failure to a header and forward the request to be handled by a different filter or service.
+If not present, the filter will send a failing response when an invalid state is reached.
+See <<derp.adoc#,DeRP Filter>> for more details.
```
Comment (dmnjohns) [inline repose-aggregator/src/docs/asciidoc/filters/valkyrie-authorization.adoc]: Good idea.

File: repose-aggregator/src/docs/asciidoc/filters/valkyrie-authorization.adoc (hunk 1 above)
Comment (adrianjgeorge) [inline repose-aggregator/src/docs/asciidoc/filters/valkyrie-authorization.adoc]: We do always go fetch if any of the features is on, but we don't always do a per device authorization. If you follow through those method calls you'll see that sometimes it just short circuits through the method.

Comment (dmnjohns) [inline repose-aggregator/src/docs/asciidoc/filters/valkyrie-authorization.adoc]: I swear I didn't see a short circuit when I was looking at the code... but I also find this code to be some of the more difficult-to-follow code in our codebase. The `ValkyrieResult` thing seems to be like pretending to have dynamic typing, and it takes me longer to figure out what data is going where. You'd think I would remember the code better by this point..

So if the feature is toggle-able, what is the configuration change to turn it on/off?

Comment (adrianjgeorge) [inline repose-aggregator/src/docs/asciidoc/filters/valkyrie-authorization.adoc]: So we'll go ask for your info if we have the device id header, the resource matches a configured one, or if you've asked for roles translation.

https://github.com/rackerlabs/repose/blob/master/repose-aggregator/components/filters/valkyrie-authorization-filter/src/main/scala/org/openrepose/filters/valkyrieauthorization/ValkyrieAuthorizationFilter.scala#L167-L168

**Comment by adrianjgeorge** [comment]
You keep having that failure in a rate limiting test. any idea what's up? Also what happened that the test reports aren't getting recorded anymore?

**Comment by dmnjohns** [comment]
I really don't know what's going on there. Let me see if I can figure it out.

**Pull request merged**